flake8_quotes/__init__.py

```diff
@@ -223,19 +233,38 @@ def get_quotes_errors(self, file_contents):
                 }
             # Otherwise (string is inline quote)
             else:
-                # If our string is a known good string, then ignore it
-                #   (')foo' -> good (continue)
-                #   "it(')s" -> good (continue)
-                #   (")foo" -> possibly bad
-                if self.config['good_single'] in unprefixed_string:
+                #   'This is a string'       -> Good
+                #   'This is a "string"'     -> Good
+                #   'This is a \"string\"'   -> Good
+                #   'This is a \'string\''   -> Bad (Q004)  Escaped inner quotes
+                #   '"This" is a \'string\'' -> Good        Changing outer quotes would not avoid escaping
+                #   "This is a string"       -> Bad (Q000)
+                #   "This is a 'string'"     -> Good        Avoids escaped inner quotes
+                #   "This is a \"string\""   -> Bad (Q000)
+                #   "\"This\" is a 'string'" -> Bad (Q000)
+                
+                middle = unprefixed_string[1:-1]
+                
+                # Correct outer quote characters
+                if last_quote_char == self.config['good_single']:
+                    if not self.config['avoid_escape']:
+                        continue
+                    if self.config['good_single'] in middle and not self.config['bad_single'] in middle:
+                        yield {
+                            'message': 'Q003 Change outer quotes to avoid escaping inner quotes',
+                            'line': start_row,
+                            'col': start_col,
+                        }
                     continue
-
-                # Output our error
-                yield {
-                    'message': 'Q000 Remove bad quotes',
-                    'line': start_row,
-                    'col': start_col,
-                }
+                
+                # Incorrect outer quote characters
+                if not self.config['good_single'] in middle or self.config['bad_single'] in middle:
```

Comment:
I don't understand the purpose of the `or self.config['bad_single'] in middle:`. Can you explain it?

Reply:
It's for situations where you have both types of quote in a string `The dev said "It's an edge case"` - Since the escape can't be avoided, I said the original preference should stand.

Reply:
I believe that's not a case we'd show an error for, is it?

Reply:
I think the introduction `avoid_escape`/better PEP8 conventions might have flipped good/bad meaning. From my perspective, `good_single` would imply that the string is okay to exist in the middle of a string

I think we have to throw out `good/bad` and move to `same/alt` since good/bad is relative to `avoid_escape`

Reply:
Request to resolve before landing

Reply:
Wrt, the original query, if the preference was for single outer quotes then
> `'The dev said "It\'s an edge case"'`
would be accepted but
> `"The dev said \"It's an edge case\""
would be rejected, with the current behaviour. The line in question is for that rejection case. Happy to change the behaviour there if you'd prefer, it's not specified explicitly in any style guide I've seen.

Wrt, Good/Bad. The current convention makes sense if you see it as referring to the outer quotes, rather than something in the string content. This interpretation also makes sense with `good_multiline`/`bad_multiline` etc where you wouldn't normally expect those values to appear in the body. (For what it's worth, I understood this straight away, but had a much harder time with the semantics elsewhere in the project!)

Reply:
Okay, glad to hear it's not a semantics issue. I don't have a lot of time so I'll keep it terse:

From what I can think of, we still don't care about `self.config['bad_single'] in middle`. We're guaranteed that the outer quote is "bad single" due to "good single" always using a `continue` in the previous conditional. The `not self.config['good_single'] in middle` should cover everything

- `"foo"`, bad and doesn't contain good single
- `"f'oo"`, good and does contain good single
- `"The dev said \"edge case\""`, bad and doesn't contain good single
- `"The dev said \"It's an edge case\""`, good and does contain good single

Reply:
The case here would be recommending `"The dev said \"It's an edge case\"" -> 'The dev said "It\'s an edge case"'` to preserve the original preference when an escape is unavoidable. However, there is a strong argument for simply allowing either case. What behaviour would you prefer?

Reply:
The primary purpose of this repo is to ensure quotes stay consistent. Avoiding escaping by switching to alternative quotes is a secondary priority so we should defer to keeping quotes consistent if we need an escape in either case (i.e. require `good_single` be used)

Reply:
I think we're in agreement then, and the code behaves as desired. The first part of the L261 (now L263) `not self.config['good_single'] in string_contents` avoids throwing an error if the string contains a good single. The second part of that line `or self.config['bad_single'] in string_contents` then *does* throw the error in the case being discussed here.

Reply:
Maybe I've over-introduced my own perspective on the goals on the package then. If what you're saying is true, then It feels like we're breaking backwards compatibility with no easy "off" switch which I'd really like to avoid

Reply:
Sorry, had a busy week. I see how this change is also a breaking change - I'll push a change to remove it.

Reply:
Cool, no worries. As a heads up, I'll be leaving for a 1 month vacation in a little over a week so I prob won't land this PR if we don't finish it by Friday April 19 (so there's time for me to revert if need be)
@zheller Might be around to land it but he hasn't dealt with this repo in a while =/